File: MiscellaneousAppKit/MiscellaneousAppKit/Configuration/Model/ConfigurationPopUpButtonDescription.m
//
//  ConfigurationPopUpButtonDescription.m
//  MiscellaneousAppKit
//
//  Created by Jinwoo Kim on 2/1/25.
//

#import "ConfigurationPopUpButtonDescription.h"

@implementation ConfigurationPopUpButtonDescription

+ (ConfigurationPopUpButtonDescription *)descriptionWithTitles:(NSArray<NSString *> *)titles selectedTitles:(NSArray<NSString *> *)selectedTitles selectedDisplayTitle:(NSString *)selectedDisplayTitle {
    return [[[ConfigurationPopUpButtonDescription alloc] initWithTitles:titles selectedTitles:selectedTitles selectedDisplayTitle:selectedDisplayTitle] autorelease];
}

- (instancetype)initWithTitles:(NSArray<NSString *> *)titles selectedTitles:(NSArray<NSString *> *)selectedTitles selectedDisplayTitle:(NSString *)selectedDisplayTitle {
    assert(![ConfigurationPopUpButtonDescription _hasDuplicatedString:titles]);
    assert(![ConfigurationPopUpButtonDescription _hasDuplicatedString:selectedTitles]);
    
    if (selectedTitles.count > 0) {
        if (selectedDisplayTitle == nil) {
            abort();
        } else if (![selectedTitles containsObject:selectedDisplayTitle]) {
            abort();
        }
        
        for (NSString *selectedTitle in selectedTitles) {
            assert(![selectedTitle isEqualToString:@"(None)"]);
        }
    } else {
        if (selectedDisplayTitle != nil) {
            abort();
        }
    }
    
    if (self = [super init]) {
        _titles = [titles copy];
        _selectedTitles = [selectedTitles copy];
        _selectedDisplayTitle = [selectedDisplayTitle copy];
    }
    
    return self;
}

- (void)dealloc {
    [_titles release];
    [_selectedTitles release];
    [_selectedDisplayTitle release];
    [super dealloc];
}

- (id)copyWithZone:(struct _NSZone *)zone {
    return [self retain];
}

+ (BOOL)_hasDuplicatedString:(NSArray<NSString *> *)strings {
    NSMutableArray<NSString *> *testArray = [[NSMutableArray alloc] initWithCapacity:strings.count];
    BOOL result = NO;
    for (NSString *string in strings) {
        if ([testArray containsObject:string]) {
            result = YES;
            break;
        }
        
        [testArray addObject:string];
    }
    [testArray release];
    return result;
}

@end

File: MiscellaneousAppKit/MiscellaneousAppKit/Configuration/Model/ConfigurationPopUpButtonDescription.swift
//
//  ConfigurationPopUpButtonDescription.swift
//  MiscellaneousAppKit
//
//  Created by Jinwoo Kim on 3/2/25.
//

import Foundation

public struct ConfigurationPopUpButtonDescription {
    public enum Title: Equatable {
        case unselected(String)
        case selected(String)
        case primarySelected(String)
        
        fileprivate var string: String {
            switch self {
            case .unselected(let string):
                return string
            case .selected(let string):
                return string
            case .primarySelected(let string):
                return string
            }
        }
    }
    
    public var titles: [Title] {
        let titles = impl.titles
        let selectedTitles = impl.selectedTitles
        let selectedDisplayTitle = impl.selectedDisplayTitle
        
        return titles
            .map { string in
                if selectedTitles.contains(string) {
                    if selectedDisplayTitle == string {
                        return .primarySelected(string)
                    } else {
                        return .selected(string)
                    }
                } else {
                    return .unselected(string)
                }
            }
    }
    
    public var selectedDisplayTitle: String? {
        impl.selectedDisplayTitle
    }
    
    private let impl: __ConfigurationPopUpButtonDescription
    
    private init(impl: __ConfigurationPopUpButtonDescription) {
        self.impl = impl
    }
    
    public init(titles: [Title]) {
        var selectedStrings: [String] = []
        var primarySelectedString: String?
        
        let strings = [String](unsafeUninitializedCapacity: titles.count) { buffer, initializedCount in
            for (index, title) in titles.enumerated() {
                buffer.baseAddress.unsafelyUnwrapped.advanced(by: index).initialize(to: title.string)
                
                switch title {
                case .unselected(_):
                    break
                case .selected(let string):
                    selectedStrings.append(string)
                case .primarySelected(let string):
                    selectedStrings.append(string)
                    assert(primarySelectedString == nil)
                    primarySelectedString = string
                }
            }
            
            initializedCount = titles.count
        }
        
        impl = __ConfigurationPopUpButtonDescription(
            titles: strings,
            selectedTitles: selectedStrings,
            selectedDisplayTitle: primarySelectedString
        )
    }
}

extension ConfigurationPopUpButtonDescription: _ObjectiveCBridgeable {
    public func _bridgeToObjectiveC() -> __ConfigurationPopUpButtonDescription {
        impl
    }
    
    public static func _forceBridgeFromObjectiveC(_ source: __ConfigurationPopUpButtonDescription, result: inout ConfigurationPopUpButtonDescription?) {
        result = ConfigurationPopUpButtonDescription(impl: source)
    }
    
    public static func _conditionallyBridgeFromObjectiveC(_ source: __ConfigurationPopUpButtonDescription, result: inout ConfigurationPopUpButtonDescription?) -> Bool {
        result = ConfigurationPopUpButtonDescription(impl: source)
        return true
    }
    
    public static func _unconditionallyBridgeFromObjectiveC(_ source: __ConfigurationPopUpButtonDescription?) -> ConfigurationPopUpButtonDescription {
        ConfigurationPopUpButtonDescription(impl: source!)
    }
}

extension ConfigurationForm {
    public struct PopUpButtonItem: Item {
        public typealias MenuTitle = ConfigurationPopUpButtonDescription.Title
        
        public let _itemModel: ConfigurationItemModel?
        public let _didTriggerAction: (Any?) -> Void
        
        public init(
            identifier: String,
            title: String,
            menuTitles: [MenuTitle],
            action: @escaping ((String) -> Void)
        ) {
            self.init(
                itemModel: ConfigurationItemModel
                    .popUpButtonItem(
                        identifier: identifier,
                        label: title,
                        valueResolver: { _ in
                            ConfigurationPopUpButtonDescription(
                                titles: menuTitles
                            )
                        }
                    ),
                didTriggerAction: { newValue in
                    action(newValue as! String)
                }
            )
        }
        
        private init(itemModel: ConfigurationItemModel, didTriggerAction: @escaping (Any?) -> Void) {
            _itemModel = itemModel
            _didTriggerAction = didTriggerAction
        }
    }
}
